### PR TITLE
HIP-584: Create Account Database Accessor

### LINKED ISSUE #5976: HIP-584: Create Account Database Accessor
### Problem

TBD

### Solution

TBD

### Alternatives

_No response_

### PR DESCRIPTION
**Pull request by stoyan-lime**
**Description**:
Create a DB accessor, that given an address gets all relevant data from relevant repositories and returns an Account.

**Related issue(s)**:

Fixes #5976

### CODE REVIEW

File: hedera-mirror-web3/src/main/java/com/hedera/mirror/web3/repository/TokenAllowanceRepository.java [hunk 1]
```diff
@@ -30,4 +31,6 @@ public interface TokenAllowanceRepository extends CrudRepository<TokenAllowance,
     @Override
     @Cacheable(cacheNames = "token_allowance", cacheManager = CACHE_MANAGER_TOKEN, unless = "#result == null")
     Optional<TokenAllowance> findById(Id id);
+
+    List<TokenAllowance> findBySpender(long spender);
```
Comment (david-bakin-sl) [inline hedera-mirror-web3/src/main/java/com/hedera/mirror/web3/repository/TokenAllowanceRepository.java:35]: I'm not sure what's going on here but given the pattern in this file and the other repositories, why doesn't this method need a sql or cache annotation?

(I'm working on files going _upward_ again so this is the first one I saw.  I see it's peers are similar.  But don't know why.)

File: hedera-mirror-web3/src/main/java/com/hedera/mirror/web3/repository/TokenAccountRepository.java [hunk 2]
```diff
@@ -21,12 +21,20 @@
 import com.hedera.mirror.common.domain.token.AbstractTokenAccount.Id;
 import com.hedera.mirror.common.domain.token.TokenAccount;
 import java.util.Optional;
+import org.jetbrains.annotations.NotNull;
```
Comment (david-bakin-sl) [inline hedera-mirror-web3/src/main/java/com/hedera/mirror/web3/repository/TokenAccountRepository.java:24]: Not lombok's notnull?

File: hedera-mirror-web3/src/main/java/com/hedera/mirror/web3/evm/store/accessor/AccountDatabaseAccessor.java [hunk 3]
```diff
@@ -0,0 +1,137 @@
...
+    @Override
+    public @NonNull Optional<Account> get(@NonNull Address address) {
+        return getEntity(address).map(this::accountFromEntity);
+    }
+
+    private Optional<Entity> getEntity(Address address) {
+        var addressBytes = address.toArrayUnsafe();
+        if (isMirror(addressBytes)) {
+            final var entityId = entityIdNumFromEvmAddress(address);
+            return entityRepository.findByIdAndDeletedIsFalse(entityId);
```
Comment (david-bakin-sl) [inline hedera-mirror-web3/src/main/java/com/hedera/mirror/web3/evm/store/accessor/AccountDatabaseAccessor.java:70]: This method name `findByIdAndDeletedIsFalse` is trying to tell you what it really does (which is nice!) but it returns an optional not a boolean ... so kind of misnamed?  Or is it just understood that Optional.empty means false?  (Which sounds reasonable.)

**Review by david-bakin-sl — COMMENTED** [review]
I like this account "aggregator" thing - hides the complexity of the actual schema to provide a single unified "account".

(Still trying to understand how the underlying cache + database works - that's just me, I'm unfamiliar with it.)

(This review doesn't include tests. I'll come back later for those and add an official checkmark.)

File: hedera-mirror-web3/src/main/java/com/hedera/mirror/web3/evm/store/accessor/AccountDatabaseAccessor.java (hunk 3 above)
Comment (stoyan-lime) [reply hedera-mirror-web3/src/main/java/com/hedera/mirror/web3/evm/store/accessor/AccountDatabaseAccessor.java:70]: The method is a standard auto-generated query. It queries the database finding matches where ID matches the parameter, and where "deleted is false". So it returns the first object matching the query, which is optional and not a boolean.
You can read more here:
https://docs.spring.io/spring-data/jpa/docs/current/reference/html/#repositories.query-methods

File: hedera-mirror-web3/src/main/java/com/hedera/mirror/web3/repository/TokenAccountRepository.java (hunk 2 above)
Comment (stoyan-lime) [reply hedera-mirror-web3/src/main/java/com/hedera/mirror/web3/repository/TokenAccountRepository.java]: Updated

File: hedera-mirror-web3/src/main/java/com/hedera/mirror/web3/repository/TokenAccountRepository.java [hunk 4]
```diff
@@ -21,12 +21,20 @@
...
 
     @Override
     @Cacheable(cacheNames = "token_account", cacheManager = CACHE_MANAGER_TOKEN, unless = "#result == null")
-    Optional<TokenAccount> findById(Id id);
+    @NonNull
+    Optional<TokenAccount> findById(@NonNull Id id);
+
+    int countByAccountIdAndAssociatedIsTrue(long accountId);
+
+    @Query("SELECT COUNT(ta) FROM TokenAccount ta WHERE ta.accountId=?1 AND balance > 0")
```
Comment (steven-sheehy) [inline hedera-mirror-web3/src/main/java/com/hedera/mirror/web3/repository/TokenAccountRepository.java:38]: For efficiency, you can combine these two with a group by:

`select count(*), balance > 0 from TokenAccount where accountId = ?1 and associated is true group by balance > 0`

Note the mirror node SQL standard is to use lowercase and spaces around operators. `ta` alias is also not needed here.

File: hedera-mirror-web3/src/test/java/com/hedera/mirror/web3/repository/NftRepositoryTest.java [hunk 5]
```diff
@@ -40,4 +40,16 @@ void findById() {
                 .returns(nft.getAccountId(), Nft::getAccountId)
                 .returns(nft.getMetadata(), Nft::getMetadata);
     }
+
+    @Test
+    void count() {
```
Comment (steven-sheehy) [inline hedera-mirror-web3/src/test/java/com/hedera/mirror/web3/repository/NftRepositoryTest.java:45]: Should match repo method.
```suggestion
    void countByAccountId() {
```

File: hedera-mirror-web3/src/test/java/com/hedera/mirror/web3/repository/TokenAccountRepositoryTest.java [hunk 6]
```diff
@@ -42,4 +42,42 @@ void findById() {
                 .returns(tokenAccount.getKycStatus(), TokenAccount::getKycStatus)
                 .returns(tokenAccount.getBalance(), TokenAccount::getBalance);
     }
+
+    @Test
+    void countByAccountIdAndAssocaition() {
```
Comment (steven-sheehy) [inline hedera-mirror-web3/src/test/java/com/hedera/mirror/web3/repository/TokenAccountRepositoryTest.java:47]:
```suggestion
    void countByAccountIdAndAssociatedIsTrue() {
```

File: hedera-mirror-web3/src/test/java/com/hedera/mirror/web3/repository/TokenAccountRepositoryTest.java [hunk 7]
```diff
@@ -42,4 +42,42 @@ void findById() {
...
+        domainBuilder
+                .tokenAccount()
+                .customize(a -> a.associated(false).accountId(accountId))
+                .persist();
+
+        assertThat(repository.countByAccountIdAndAssociatedIsTrue(accountId)).isEqualTo(2);
+    }
+
+    @Test
+    void countByAccountIdAndBalance() {
```
Comment (steven-sheehy) [inline hedera-mirror-web3/src/test/java/com/hedera/mirror/web3/repository/TokenAccountRepositoryTest.java:66]:
```suggestion
    void countByAccountIdAndPositiveBalance() {
```

File: hedera-mirror-web3/src/main/java/com/hedera/mirror/web3/evm/store/accessor/AccountDatabaseAccessor.java [hunk 8]
```diff
@@ -0,0 +1,137 @@
...
+                                EntityNum.fromLong(tokenAllowance.getSpender())),
+                        AbstractTokenAllowance::getAmount,
+                        NO_DUPLICATE_MERGE_FUNCTION,
+                        TreeMap::new));
+    }
+
+    private SortedSet<FcTokenAllowanceId> getApproveForAllNfts(Long spenderId) {
+        return nftAllowanceRepository.findBySpenderAndApprovedForAllIsTrue(spenderId).stream()
+                .map(nftAllowance -> new FcTokenAllowanceId(
+                        EntityNum.fromLong(nftAllowance.getTokenId()), EntityNum.fromLong(nftAllowance.getSpender())))
```
Comment (steven-sheehy) [inline hedera-mirror-web3/src/main/java/com/hedera/mirror/web3/evm/store/accessor/AccountDatabaseAccessor.java:126]: Use EntityId

**Review by steven-sheehy — CHANGES_REQUESTED** [review]

File: hedera-mirror-web3/src/main/java/com/hedera/mirror/web3/evm/store/accessor/AccountDatabaseAccessor.java [hunk 9]
```diff
@@ -0,0 +1,137 @@
...
+        } else {
+            return entityRepository.findByEvmAddressAndDeletedIsFalse(addressBytes);
+        }
+    }
+
+    private Account accountFromEntity(Entity entity) {
+        EntityId entityId = entity.toEntityId();
+        return new Account(
+                idFromEntityId(entityId),
+                entity.getExpirationTimestamp(),
```
[the remark this answers is not in the document]
Comment (stoyan-lime) [reply hedera-mirror-web3/src/main/java/com/hedera/mirror/web3/evm/store/accessor/AccountDatabaseAccessor.java]: Added fallback.

File: hedera-mirror-web3/src/main/java/com/hedera/mirror/web3/evm/store/accessor/AccountDatabaseAccessor.java [hunk 10]
```diff
@@ -0,0 +1,137 @@
...
+                entity.getBalance(),
+                entity.getDeleted(),
```
[the remark this answers is not in the document]
Comment (stoyan-lime) [reply hedera-mirror-web3/src/main/java/com/hedera/mirror/web3/evm/store/accessor/AccountDatabaseAccessor.java]: Added fallback.

File: hedera-mirror-web3/src/main/java/com/hedera/mirror/web3/evm/store/accessor/AccountDatabaseAccessor.java [hunk 11]
```diff
@@ -0,0 +1,137 @@
...
+            return entityRepository.findByIdAndDeletedIsFalse(entityId);
+        } else {
+            return entityRepository.findByEvmAddressAndDeletedIsFalse(addressBytes);
+        }
+    }
+
+    private Account accountFromEntity(Entity entity) {
+        EntityId entityId = entity.toEntityId();
+        return new Account(
+                idFromEntityId(entityId),
```
[the remark this answers is not in the document]
Comment (stoyan-lime) [reply hedera-mirror-web3/src/main/java/com/hedera/mirror/web3/evm/store/accessor/AccountDatabaseAccessor.java]: Updated.

File: hedera-mirror-web3/src/main/java/com/hedera/mirror/web3/evm/store/accessor/AccountDatabaseAccessor.java [hunk 12]
```diff
@@ -0,0 +1,137 @@
...
+                entity.getExpirationTimestamp(),
+                entity.getBalance(),
+                entity.getDeleted(),
+                getOwnedNfts(entityId),
```
[the remark this answers is not in the document]
Comment (stoyan-lime) [reply hedera-mirror-web3/src/main/java/com/hedera/mirror/web3/evm/store/accessor/AccountDatabaseAccessor.java]: Updated.

File: hedera-mirror-web3/src/main/java/com/hedera/mirror/web3/evm/store/accessor/AccountDatabaseAccessor.java [hunk 13]
```diff
@@ -0,0 +1,137 @@
...
+import java.util.TreeSet;
+import java.util.function.BinaryOperator;
+import java.util.stream.Collectors;
+import lombok.NonNull;
+import lombok.RequiredArgsConstructor;
+import org.hyperledger.besu.datatypes.Address;
+import org.springframework.beans.factory.annotation.Autowired;
+
+@RequiredArgsConstructor(onConstructor = @__(@Autowired))
+public class AccountDatabaseAccessor extends DatabaseAccessor<Address, Account> {
```
[the remark this answers is not in the document]
Comment (stoyan-lime) [reply hedera-mirror-web3/src/main/java/com/hedera/mirror/web3/evm/store/accessor/AccountDatabaseAccessor.java:48]: Updated.

File: hedera-mirror-web3/src/main/java/com/hedera/mirror/web3/evm/store/accessor/AccountDatabaseAccessor.java [hunk 14]
```diff
@@ -0,0 +1,137 @@
...
+    private Id idFromEntityId(EntityId entityId) {
+        return new Id(entityId.getShardNum(), entityId.getRealmNum(), entityId.getEntityNum());
+    }
+
+    private long getOwnedNfts(EntityId accountId) {
+        return nftRepository.countByAccountId(accountId);
+    }
+
+    private SortedMap<EntityNum, Long> getCryptoAllowances(Long spenderId) {
+        return cryptoAllowanceRepository.findBySpender(spenderId).stream()
```
[the remark this answers is not in the document]
Comment (stoyan-lime) [reply hedera-mirror-web3/src/main/java/com/hedera/mirror/web3/evm/store/accessor/AccountDatabaseAccessor.java]: Fixed.

File: hedera-mirror-web3/src/main/java/com/hedera/mirror/web3/evm/store/accessor/AccountDatabaseAccessor.java (hunk 8 above)
Comment (stoyan-lime) [reply hedera-mirror-web3/src/main/java/com/hedera/mirror/web3/evm/store/accessor/AccountDatabaseAccessor.java]: Updated.

File: hedera-mirror-web3/src/main/java/com/hedera/mirror/web3/evm/store/accessor/AccountDatabaseAccessor.java [hunk 15]
```diff
@@ -0,0 +1,137 @@
...
+                .collect(Collectors.toMap(
+                        cryptoAllowance -> EntityNum.fromLong(cryptoAllowance.getOwner()),
```
[the remark this answers is not in the document]
Comment (stoyan-lime) [reply hedera-mirror-web3/src/main/java/com/hedera/mirror/web3/evm/store/accessor/AccountDatabaseAccessor.java]: Updated.

File: hedera-mirror-web3/src/main/java/com/hedera/mirror/web3/repository/TokenAccountRepository.java [hunk 16]
```diff
@@ -21,12 +21,20 @@
...
+    int countByAccountIdAndPositiveBalance(long accountId);
```
[the remark this answers is not in the document]
Comment (stoyan-lime) [reply hedera-mirror-web3/src/main/java/com/hedera/mirror/web3/repository/TokenAccountRepository.java:37]: Updated.

File: hedera-mirror-web3/src/main/java/com/hedera/mirror/web3/repository/NftRepository.java [hunk 17]
```diff
@@ -29,4 +30,6 @@ public interface NftRepository extends CrudRepository<Nft, NftId> {
     @Override
     @Cacheable(cacheNames = "nft", cacheManager = CACHE_MANAGER_TOKEN, unless = "#result == null")
     Optional<Nft> findById(NftId nftId);
+
+    long countByAccountId(EntityId accountId);
```
[the remark this answers is not in the document]
Comment (stoyan-lime) [reply hedera-mirror-web3/src/main/java/com/hedera/mirror/web3/repository/NftRepository.java]: How important is it to check for the entity.deleted property?
To check for it we either have to define a relation between the Nft and Entity table
or create a native query, which is problematic as converters are used for account ids.

File: hedera-mirror-web3/src/main/java/com/hedera/mirror/web3/repository/TokenAccountRepository.java (hunk 4 above)
Comment (steven-sheehy) [reply hedera-mirror-web3/src/main/java/com/hedera/mirror/web3/repository/TokenAccountRepository.java]: You should switch to native query and adjust table/columns appropriately

File: hedera-mirror-web3/src/main/java/com/hedera/mirror/web3/repository/NftRepository.java (hunk 17 above)
[the remark this answers is not in the document]
Comment (steven-sheehy) [reply hedera-mirror-web3/src/main/java/com/hedera/mirror/web3/repository/NftRepository.java]: Very important. Just use long parameters and define native query

[the remark this answers is not in the document]
Comment (stoyan-lime) [reply hedera-mirror-web3/src/main/java/com/hedera/mirror/web3/repository/NftRepository.java]: Updated.

**Review by steven-sheehy — CHANGES_REQUESTED** [review]

**Comment by steven-sheehy** [comment]
2 code smells to address

File: hedera-mirror-web3/src/main/java/com/hedera/mirror/web3/evm/store/accessor/AccountDatabaseAccessor.java [hunk 18]
```diff
@@ -0,0 +1,149 @@
...
+    private Long getExpiration(Entity entity) {
+        if (entity.getExpirationTimestamp() != null) {
+            return entity.getExpirationTimestamp();
+        }
+
+        if (entity.getCreatedTimestamp() != null && entity.getAutoRenewPeriod() != null) {
+            return entity.getCreatedTimestamp() + entity.getAutoRenewPeriod();
+        }
+
+        return DEFAULT_EXPIRY_TIMESTAMP;
```
Comment (xin-hedera) [inline hedera-mirror-web3/src/main/java/com/hedera/mirror/web3/evm/store/accessor/AccountDatabaseAccessor.java:96]: the timestamps are in different resolution

- nanosecond: entity.createdTimestamp and entity.expirationTimestamp
- second: entity.autoRenewPeriod
- microsecond: DEFAULT_EXPIRY_TIMESTAMP

**Review by xin-hedera — CHANGES_REQUESTED** [review]

File: hedera-mirror-web3/src/main/java/com/hedera/mirror/web3/evm/store/accessor/AccountDatabaseAccessor.java [hunk 19]
```diff
@@ -0,0 +1,149 @@
...
+    }
+
+    private Account accountFromEntity(Entity entity) {
+        return new Account(
+                new Id(entity.getShard(), entity.getRealm(), entity.getNum()),
+                getExpiration(entity),
+                Optional.ofNullable(entity.getBalance()).orElse(0L),
+                Optional.ofNullable(entity.getDeleted()).orElse(false),
+                getOwnedNfts(entity.toEntityId()),
+                Optional.ofNullable(entity.getAutoRenewPeriod()).orElse(Long.MAX_VALUE),
```
[the remark this answers is not in the document]
Comment (stoyan-lime) [reply hedera-mirror-web3/src/main/java/com/hedera/mirror/web3/evm/store/accessor/AccountDatabaseAccessor.java]: Updated.

File: hedera-mirror-web3/src/main/java/com/hedera/mirror/web3/evm/store/accessor/AccountDatabaseAccessor.java [hunk 20]
```diff
@@ -0,0 +1,149 @@
...
+                idFromEntityId(entity.getProxyAccountId()),
+                entity.getMaxAutomaticTokenAssociations(),
```
[the remark this answers is not in the document]
Comment (stoyan-lime) [reply hedera-mirror-web3/src/main/java/com/hedera/mirror/web3/evm/store/accessor/AccountDatabaseAccessor.java]: Updated.

File: hedera-mirror-web3/src/test/java/com/hedera/mirror/web3/repository/TokenAllowanceRepositoryTest.java [hunk 21]
```diff
@@ -38,4 +38,13 @@ void findAllowance() {
                         .orElse(0L))
                 .isEqualTo(tokenAllowance.getAmount());
     }
+
+    @Test
+    void findBySpender() {
```
[the remark this answers is not in the document]
Comment (stoyan-lime) [reply hedera-mirror-web3/src/test/java/com/hedera/mirror/web3/repository/TokenAllowanceRepositoryTest.java]: Updated.

File: hedera-mirror-web3/src/main/java/com/hedera/mirror/web3/repository/TokenAccountRepository.java (hunk 4 above)
Comment (stoyan-lime) [reply hedera-mirror-web3/src/main/java/com/hedera/mirror/web3/repository/TokenAccountRepository.java]: Updated.

File: hedera-mirror-web3/src/main/java/com/hedera/mirror/web3/evm/store/accessor/AccountDatabaseAccessor.java (hunk 18 above)
Comment (stoyan-lime) [reply hedera-mirror-web3/src/main/java/com/hedera/mirror/web3/evm/store/accessor/AccountDatabaseAccessor.java]: Updated.
For future reference where can I see the resolutions of different timestamps?

Comment (xin-hedera) [reply hedera-mirror-web3/src/main/java/com/hedera/mirror/web3/evm/store/accessor/AccountDatabaseAccessor.java]: there's no document in code for these fields. You can check what values are saved to the fields in CryptoCreateTransactionHandler/CryptoUpdateTransactionHandler.

**Review by steven-sheehy — APPROVED** [review]
LGTM

**Pull request merged**